Implement on-the-fly compression

philippotto opened the pull request:
Also see #20.

In discussion with @normanrz, we decided that unbuffered "live compression" for write calls would already be a very valuable feature. This means that blocks are compressed and written directly "as they come" in morton order. Write-calls at random positions are forbidden, for that reason. Additionally, this means that we don't change the semantics of the `write` call, since adding a buffer would delay the "actual writing" to a later point in time.

The API interface didn't change. However, it is now possible to `create` a dataset with a LZ4 header and make `write` calls (previously, this produced an "unsupported" message). Have a look at the new tests to see the usage.

@amotta My rust experience is quite limited. So, feel free to give feedback :)

amotta commented:
Hi @philippotto 

Thanks a lot for the pull request. I like the idea of on-the-fly compression and agree that it will be a very helpful feature.

Unfortunately, the current implementation is not yet quite strict enough. It allowed me to perform write operations, which lead to WKW files with invalid headers (i.e., entries in the jump table). This in turn leads to a crash of the reader (that's a robustness issue that should be addressed another time).

The problem is that I'm allowed to skip blocks when using the on-the-fly compression. Consider the example below:

```python
import numpy as np
import wkw

dataset = wkw.Dataset.create('tests/tmp', wkw.Header(np.uint8, block_type=wkw.Header.BLOCK_TYPE_LZ4))

test_data = np.random.uniform(0, 255, (64, 64, 64)).astype(np.uint8)
dataset.write((0, 0, 0), test_data)

test_data = np.random.uniform(0, 255, (64, 128, 64)).astype(np.uint8)
dataset.write((64, 0, 0), test_data)

dataset.read((0, 0, 0), (128, 128, 128))
```

The first `write` fills blocks 1..8. The second `write`, however, fills blocks 9..16 and 25..32. Importantly, blocks 17..24 are skipped:

![bug-illustration](https://user-images.githubusercontent.com/256037/46247396-2341cf80-c40b-11e8-8736-59e029e0d009.png)

As a result, the jump table entries for the skipped blocks are empty (i.e., zero):
```
00000000  57 4b 57 01 55 02 01 01  10 00 04 00 00 00 00 00  |WKW.U...........|
00000010  92 80 04 00 00 00 00 00  14 01 05 00 00 00 00 00  |................| Blocks 1, 2, ..
00000020  96 81 05 00 00 00 00 00  18 02 06 00 00 00 00 00  |................|
00000030  9a 82 06 00 00 00 00 00  1c 03 07 00 00 00 00 00  |................|
00000040  9e 83 07 00 00 00 00 00  20 04 08 00 00 00 00 00  |........ .......| .., 7, 8 are filled in first write.
00000050  a2 84 08 00 00 00 00 00  24 05 09 00 00 00 00 00  |........$.......| Blocks 9, 10, ..
00000060  a6 85 09 00 00 00 00 00  28 06 0a 00 00 00 00 00  |........(.......|
00000070  aa 86 0a 00 00 00 00 00  2c 07 0b 00 00 00 00 00  |........,.......|
00000080  ae 87 0b 00 00 00 00 00  30 08 0c 00 00 00 00 00  |........0.......| .., 15, 16 are filled in second write.
00000090  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................| Blocks 17, 18, ..
000000a0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000c0  00 00 00 00 00 00 00 00  30 08 0c 00 00 00 00 00  |........0.......| .., 23, 24 are skipped.
000000d0  b2 88 0c 00 00 00 00 00  34 09 0d 00 00 00 00 00  |........4.......| Blocks 25, 26, ..
000000e0  b6 89 0d 00 00 00 00 00  38 0a 0e 00 00 00 00 00  |........8.......|
000000f0  ba 8a 0e 00 00 00 00 00  3c 0b 0f 00 00 00 00 00  |........<.......|
00000100  be 8b 0f 00 00 00 00 00  40 0c 10 00 00 00 00 00  |........@.......| .., 31, 32 are filled in second write.
```
The reader then calculates the length of block 17 (see `Header::block_size_on_disk`) as `0u64 - 788528u64 which is 18446744073708763088u64`. Obviously, this doesn't make sense and then leads to the following crash:
```
thread '<unnamed>' panicked at 'index 18446744073708763088 out of range for slice of length 32912', libcore/slice/mod.rs:785:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
fatal runtime error: failed to initiate panic, error 5
[1]    4414 abort (core dumped)  ipython
```

To fix this bug, we should make sure that on-the-fly writes begin at the first empty block and that the writes are contiguous in morton order (i.e., no skipped blocks).

philippotto commented:
Thanks for your very detailed write-up, Alessandro!

Enforcing continuous morton-block order is certainly a solution to this problem. However, I'm wondering how much this would diminish the value of on-the-fly compression. Additionally, wouldn't we need to enforce that a file is also written "until the end"? Otherwise, you could write blocks 10 to 100 and if trying to read block 101, it would crash, too?

What do you think about an alternative solution which would detect such "missing blocks" and return empty data in these cases? Then, it would behave more closely to the already existing compression mechanism which also supports compressing empty blocks.

For this solution, we wouldn't need to adapt the write process. Instead, during reading we could check whether the jump table contains valid values for the requested block. I.e., if the jump table contains zeros at the desired position, we know that no block was written for that position, which is why we can return an empty block.

What do you think, @amotta @normanrz?

amotta commented:
Hi Philipp. Thanks for your input.

I'll have to think a little bit about how to best handle missing blocks in LZ4 compressed WKW files.

For now (and to implement what you guys originally set out to do), I'd suggest that we add your code for on-the-fly compression, but limit it to writing whole WKW files. This way, we don't have to worry about missing blocks and simultaneously cover the most common use case.

philippotto commented:
Alright, that sounds like a plan. I pushed an update which enforces to write complete blocks when compression is turned on. So, the PR might be should to get merged.

I also pushed another branch (`sparse-compression-fix`) where empty blocks are returned when the corresponding jump table values are zero. However, the tests don't pass yet (not sure why; but I'd wait with debugging this until we made a final decision on how to tackle this conceptually).

amotta commented on rust/src/dataset.rs:
```diff
@@ -147,6 +140,13 @@ impl Dataset {
                         // offsets
                         let cur_src_pos = cur_box.min() - dst_pos;
                         let cur_dst_pos = cur_box.min() - cur_file_box.min();
+                        if self.header.block_type != BlockType::Raw {  
```
Please use `self.header.block_type == BlockType::LZ4 || self.header.block_type == BlockType::LZ4HC`` instead. This is correct at the moment, but might cause trouble as soon as we add a new non-LZ4 block type.

amotta commented on rust/src/dataset.rs:
```diff
@@ -147,6 +140,13 @@ impl Dataset {
                         let cur_src_pos = cur_box.min() - dst_pos;
                         let cur_dst_pos = cur_box.min() - cur_file_box.min();
+                        if self.header.block_type != BlockType::Raw {  
+                            if cur_box != cur_file_box {
```
If I try to write a (1025, 1024, 1024) data volume to location (0, 0, 0), this will write the first (1024)³ cube and subsequently throw an error for the (1, 1024, 1024) plane at (1024, 0, 0). This partial write is problematic because it leaves the dataset in an inconsistent state.

Please check that `dst_pos` and `mat.shape` are zero modulo the file volume at the start of the function.

amotta commented on rust/src/file.rs:
```diff
@@ -153,7 +152,19 @@ impl File {
-        let iter = Iter::new(self.header.file_len_log2 as u32, dst_box_boxes)?;
+        let mut iter = Iter::new(self.header.file_len_log2 as u32, dst_box_boxes)?.peekable();
+
+        if self.header.block_type != BlockType::Raw {
```
This is not needed for the whole-file case. Please leaves this complexity out for this pull request.

amotta commented on rust/src/file.rs:
```diff
@@ -176,11 +188,24 @@ impl File {
             // fill / modify buffer
             buf_mat.copy_from(cur_dst_pos, src_mat, cur_src_box)?;
 
+            if self.header.block_type == BlockType::Raw {
```
Hmm, not sure whether this case distinction is necessary. The `seek_block` method doesn't do any seeking if we're already at the right location. Ideally, we keep it simple and call `seek_block` no matter what.

philippotto replied:
Good point! Nevertheless, we have to set `self.block_idx` conditionally here, since the seek method would otherwise try to calculate the correct position via the jump table which is not complete yet. I left the `self.block_idx = ...` line in and moved the seek method out of the if, since it will be a no-op if block_idx is already correct.

amotta commented on rust/src/file.rs:
```diff
@@ -354,4 +386,28 @@ impl File {
         }
     }
+
+    fn seek_to_first_empty_block(&mut self) -> Result<u64> {
```
Not needed for the whole-file case.

amotta commented on rust/src/file.rs:
```diff
@@ -312,6 +337,13 @@ impl File {
         let block_idx = self.block_idx.unwrap();
         let jump_table = &mut *self.header.jump_table.as_mut().unwrap();
         jump_table[block_idx as usize] = jump_entry;
+        if block_idx > 0 {
```
This is not needed for the whole-file case.

philippotto commented:
> Thanks! I'd suggest that we avoid any unnecessary complexity by limiting the code to the bare minimum that is needed to write entire LZ4 compressed files.

Sure, sounds reasonable. Thanks for your PR feedback! I incorporated your suggestions and also adapted the tests. Let me know if this ready to be merged!

amotta commented on rust/src/file.rs:
```diff
@@ -176,11 +175,22 @@ impl File {
             buf_mat.copy_from(cur_dst_pos, src_mat, cur_src_box)?;
 
-            // write data
+            if self.header.block_type == BlockType::LZ4 || self.header.block_type == BlockType::LZ4HC {
```
I'm nearly 99 % sure that this if-block is not needed :)

The `self.block_idx` variable is set by the call to `Dataset::write_block` a few lines later. So, the if-block only every has an effect (i.e., avoiding a seek) for the very first block. It's true that this statement avoids a (currently) unnecessary seek for the very first block. Still, to reduce complexity, I'd kick it out.

philippotto replied:
You're correct! In my mind, I was still thinking of the "sparse write" case :)

amotta commented on rust/src/file.rs:
```diff
@@ -176,11 +175,22 @@ impl File {
             self.write_block(buf_mat.as_slice())?;
         }
 
+        if self.header.block_type != BlockType::Raw {
```
Please compare against LZ4 and LZ4HC, as above.

philippotto commented:
> It's possible that the new content is better compressible than what had been written to the file previously. In this situation, the new data stream will be shorter than the existing file. To avoid having left-over data at the end of the file, it should be truncated after writing the last block.

Very valid point! I added another test case which covers this case and implemented the necessary changes to make it pass. Let me know in case something else is missing!

philippotto commented on rust/src/file.rs:
```diff
@@ -212,18 +218,24 @@ impl File {
-            _ => Ok(())
+            BlockType::LZ4 | BlockType::LZ4HC => {
+                let last_block_idx = self.header.file_vol() - 1;
+                let jump_table = &mut *self.header.jump_table.as_mut().unwrap();
```
> In fact, this whole function doesn't modify the wkw::File at all. If you want, you could even improve the code by changing the function signature to fn truncate(&self).

 Ah, interesting, I assumed that a `mut` was necessary since `set_len` mutates the file, but I'm not fully acquainted with the borrow model yet, so that was a helpful hint, thanks!

philippotto commented:
Right back at you! Thanks for your helpful feedback. I incorporated the last changes. I'll probably wait with merging until Monday, though, since I want to clarify some things with norman first.

philippotto commented:
Sorry, I actually wanted to squash the PR, but I'm used to "squash" being selected by default. This was not the case for me, though. We could probably rebase and force push, but I'm not sure whether that's worth the hassle.

@amotta I'd like to bump the version number for wkw, but I'm not sure whether the following steps are correct.

I'd commit and push the following commit directly to master. Is this the right way to do that?

```diff
diff --git a/c/Cargo.toml b/c/Cargo.toml
index f2f1efd..356de1c 100644
--- a/c/Cargo.toml
+++ b/c/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "wkw"
-version = "0.1.0"
+version = "0.1.1"
 authors = ["Alessandro Motta <alessandro.motta@brain.mpg.de>"]
 
 [dependencies]
diff --git a/matlab/rust/wkw_compress/Cargo.toml b/matlab/rust/wkw_compress/Cargo.toml
index 9d78e6f..03b0b4d 100644
--- a/matlab/rust/wkw_compress/Cargo.toml
+++ b/matlab/rust/wkw_compress/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "wkw_compress"
-version = "1.4.0"
+version = "1.5.0"
 authors = ["Alessandro Motta <alessandro.motta@brain.mpg.de>"]
 build = "build.rs"
 
diff --git a/matlab/rust/wkw_init/Cargo.toml b/matlab/rust/wkw_init/Cargo.toml
index ea74c9f..6e73f9b 100644
--- a/matlab/rust/wkw_init/Cargo.toml
+++ b/matlab/rust/wkw_init/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "wkw_init"
-version = "1.4.0"
+version = "1.5.0"
 authors = ["Alessandro Motta <alessandro.motta@brain.mpg.de>"]
 build = "build.rs"
 
diff --git a/matlab/rust/wkw_load/Cargo.toml b/matlab/rust/wkw_load/Cargo.toml
index 3244453..5f3215f 100644
--- a/matlab/rust/wkw_load/Cargo.toml
+++ b/matlab/rust/wkw_load/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "wkw_load"
-version = "1.4.0"
+version = "1.5.0"
 authors = ["Alessandro Motta <alessandro.motta@brain.mpg.de>"]
 build = "build.rs"
 
diff --git a/matlab/rust/wkw_mex/Cargo.toml b/matlab/rust/wkw_mex/Cargo.toml
index bb60b4d..c483a2b 100644
--- a/matlab/rust/wkw_mex/Cargo.toml
+++ b/matlab/rust/wkw_mex/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "wkw_mex"
-version = "1.4.0"
+version = "1.5.0"
 authors = ["Alessandro Motta <alessandro.motta@brain.mpg.de>"]
 
 [dependencies]
diff --git a/matlab/rust/wkw_save/Cargo.toml b/matlab/rust/wkw_save/Cargo.toml
index 35622fc..e1e692c 100644
--- a/matlab/rust/wkw_save/Cargo.toml
+++ b/matlab/rust/wkw_save/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "wkw_save"
-version = "1.4.0"
+version = "1.5.0"
 authors = ["Alessandro Motta <alessandro.motta@brain.mpg.de>"]
 build = "build.rs"
 
diff --git a/python/setup.py b/python/setup.py
index 2e15a1d..25bd5ab 100644
--- a/python/setup.py
+++ b/python/setup.py
@@ -44,7 +44,7 @@ class BuildPyCommand(build_py):
 
 setup(
     name="wkw",
-    version="0.0.5",
+    version="0.0.6",
     author="Alessandro Motta",
     author_email="alessandro.motta@brain.mpg.de",
     url="https://github.com/scalableminds/webknossos-wrap",
diff --git a/rust/Cargo.toml b/rust/Cargo.toml
index 625fffe..ef03130 100644
--- a/rust/Cargo.toml
+++ b/rust/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "wkwrap"
-version = "1.4.0"
+version = "1.5.0"
 authors = ["Alessandro Motta <alessandro.motta@brain.mpg.de>"]
 description = "webKNOSSOS wrapper is a file format designed for large-scale, three-dimensional voxel datasets. It was optimized for high-speed access to data subvolumes, and supports multi-channel data and dataset compression."
 repository = "https://github.com/scalableminds/webknossos-wrap"

amotta commented:
Hi @philippotto 

Bumping the version numbers for the Rust, C, and Python components seems fine to me. For now, you can leave out the MATLAB-related stuff.

`wkwInit` (the equivalence of `wkw::Dataset::create`) doesn't allow yet the creation of LZ4 compressed datasets. I'd like to fix this before compiling and distributing a new version.
